digitalstrom_mqtt/mqtt.go

```diff
@@ -194,6 +224,254 @@ func (dm *DigitalstromMqtt) getTopic(deviceType string, deviceId string, deviceN
+				device.Name,
+				device.OutputChannels[1],
+				"command")
+			message["tilt_status_template"] = "{{ value | int }}"
+		}
+	} else {
+		// raise error
+	}
+	json, err := json.Marshal(message)
+	if err != nil {
+		return nil, err
+	}
+	return []HassDiscoveryMessage{
+		{
+			topic:   topic,
+			message: json,
+		},
+	}, nil
+}
+
+// Return the definition of the power and energy sensors coming from the circuit
+// devices.
+func (dm *DigitalstromMqtt) circuitToHomeAssistantDiscoveryMessage(circuit digitalstrom.Circuit) ([]HassDiscoveryMessage, error) {
+	device_config := map[string]interface{}{
+		"configuration_url": "https://" + dm.ds_config.Host,
+		"identifiers":       []interface{}{circuit.Dsid},
+		"manufacturer":      "DigitalStrom",
+		"model":             circuit.HwName,
+		"name":              circuit.Name,
+	}
+	availability := []interface{}{
+		map[string]interface{}{
+			"topic":                 dm.getStatusTopic(),
+			"payload_available":     Online,
+			"payload_not_available": Offline,
+		},
+	}
+	// Setup configuration for a MQTT Cover in Home Assistant:
+	// https://www.home-assistant.io/integrations/sensor.mqtt/
+	// Define sensor for power consumption. This is a straightforward
+	// definition.
+	power_topic := dm.config.HomeAssistantDiscoveryPrefix + "/sensor/" + circuit.Dsid + "/power/config"
+	power_message := map[string]interface{}{
+		"device":            device_config,
+		"name":              "Power " + circuit.Name,
+		"unique_id":         circuit.Dsid + "_power",
+		"retain":            dm.config.Retain,
+		"availability":      availability,
+		"availability_mode": "all",
+		"state_topic": dm.getTopic(
+			"circuits",
+			circuit.Dsid,
+			circuit.Name,
+			"consumptionW",
+			"state"),
+		"unit_of_measurement": "W",
+		"device_class":        "power",
+		"icon":                "mdi:flash",
+		"qos":                 0,
+	}
+	// Define the energy sensor. We need to define the state class in order to
+	// make sure statistics are bing computed and stored. We also use the
+	// `value_template` field to make the conversion from Ws reported in the
+	// MQTT topic, to kWh which is the default unit of measurement of energy in
+	// Home Assistant.
+	energy_topic := dm.config.HomeAssistantDiscoveryPrefix + "/sensor/" + circuit.Dsid + "/energy/config"
+	energy_message := map[string]interface{}{
+		"device":            device_config,
+		"name":              "Energy " + circuit.Name,
+		"unique_id":         circuit.Dsid + "_energy",
+		"retain":            dm.config.Retain,
+		"availability":      availability,
+		"availability_mode": "all",
+		"state_topic": dm.getTopic(
+			"circuits",
+			circuit.Dsid,
+			circuit.Name,
+			"EnergyWs",
+			"state"),
+		"unit_of_measurement": "kWh",
+		"device_class":        "energy",
+		"state_class":         "total_increasing",
+		// Convert the vaue from Ws to kWh which is the default energy unit in
+		// Home Assistant
+		"value_template": "{{ value | float / (3600*1000) | round(3) }}",
```

Comment:
The `round(3)` is definitively not working for me.... I think we can have only 1 digit (or none). 3 digits are useless IMO.
![Screenshot from 2022-03-18 09-22-08](https://user-images.githubusercontent.com/3659220/158963511-84b0d054-e565-421c-a314-36801814a0d0.png)

Reply:
Sorry, there was a problem in template, now it should be working. Keeping three decimal points I believe is useful to have resolution up to Wh.

<img width="1853" alt="Screenshot 2022-03-19 at 10 23 56" src="https://user-images.githubusercontent.com/4179508/159115518-e5fc5d07-2311-413c-984c-52926886bd19.png">
g

Reply:
thx